### ISSUE TITLE
Problem with expanding Scope in Uber API

### ISSUE THREAD

**Issue by DataWookie** [hadley/httr#379]
Hi Hadley,

I trust that you are having a good time in Stanford. Wish I could be at useR2016 too!

I've been having some trouble accessing the full scope of the Uber API via the httr package.

First let's start with that works:

```
endpoint <- httr::oauth_endpoint(
  authorize = "https://login.uber.com/oauth/v2/authorize",
  access    = "https://login.uber.com/oauth/v2/token"
)
app <- httr::oauth_app("uber", key = client_key, secret = client_secret)

scope = "profile"
token <- httr::oauth2.0_token(endpoint, app, scope = scope, cache = FALSE)
```

That executes fine and I'm left with an authentication token with which I am able to access a subset of the Uber API. However, to get at all of the functionality I need to specify a broader range of options for scope.

So I make the change
```
scope = "profile,request"
```
but that results in the following error:
```
Waiting for authentication in browser...
Press Esc/Ctrl + C to abort
Authentication complete.
Error in init_oauth2.0(self$endpoint, self$app, scope = self$params$scope,  : 
  Bad Request (HTTP 400). Failed to get an access token.
```

In the browser eveything seems to have proceeded as before.

I did read somewhere that https was required for the redirect URL to make this work. Could that perhaps be the source of the problem?

Please let me know if I can provide additional information to resolve this issue.

Best regards,
Andrew.

**Comment by DataWookie** [comment]
> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04 LTS

locale:
 [1] LC_CTYPE=en_ZA.UTF-8       LC_NUMERIC=C               LC_TIME=en_ZA.UTF-8        LC_COLLATE=en_ZA.UTF-8    
 [5] LC_MONETARY=en_ZA.UTF-8    LC_MESSAGES=en_ZA.UTF-8    LC_PAPER=en_ZA.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_ZA.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] httr_1.2.0.9000

loaded via a namespace (and not attached):
[1] R6_2.1.2        rsconnect_0.4.3 tools_3.3.1

**Comment by DataWookie** [comment]
I see that people are having similar issues with Rfacebook (https://github.com/pablobarbera/Rfacebook/issues/77).

**Comment by Stormseer** [comment]
Hey! I figured I'd cross-post what I found while poking around for a potential fix for the Rfacebook issue.

> I'm also getting the same issue, so I decided to poke around a little and I found the following:
> 
> > **All Callback Subscriptions Must Use HTTPS** - When we launched Graph API v2.5 last October we began requiring new Webhooks (formerly known as Real Time Updates) subscriptions to be configured for HTTPS. With v2.6, this is now required for all new Dynamic Pricing and Deauthorization endpoints as well. Beginning 90 days from today, we will require all callback subscriptions to be configured for HTTPS. At that point, we will stop sending updates to non-HTTPS endpoints, regardless of when they were created. This applies to Webhooks, Webhooks for Payments, Dynamic Pricing, and Deauthorization. You can update your subscriptions on your app dashboard.
> 
> [(Source)](https://developers.facebook.com/docs/apps/changelog)
> 
> I believe this issue is caused by Rfacebook using HTTP, which is no longer supported.

The error being experienced is due to the API(s) denying all HTTP requests, requiring HTTPS instead.

**Comment by DataWookie** [comment]
Hi Stormseer, yes, I was aware of this. See my comment just before yours. I guess this issue then becomes whether it will be possible to access the APIs with HTTPS requirement using the httr package?

**Comment by hadley** [comment]
Given that you're supplying https urls, and that it works with default scope, I suspect your problem lies with the specification of multlple scopes, not something more esoteric.  Similarly, I don't see any evidence that Rfacebook uses http, so I think you need to do more exploration.  

Generally, the best place to start is to get more information about the requests and responses by wrapping the code in `httr::with_verbose()`.

**Comment by hadley** [comment]
You might also want to try with the previous version of httr to confirm that it's not a regression.

**Comment by lyons422** [comment]
I experienced the same HTTP 400 issue when trying to connect with facebook. I ran the same code by using httr 1.1 and httr 1.2 respectively. It turns out 1.1 version has no problem with the authentication process. Guess there's something wrong with the code of new 1.2 version?

**Comment by hadley** [comment]
It works for me locally

**Comment by DataWookie** [comment]
Hi Hadley,

Thanks for the suggestions.

1. I tried reverting to 1.1.0 but ran into the same problem.
2. I wrapped my authentication call in with_verbose() [did not know that functionality existed!] and found something interesting:

(a) the successful authentication using only "profile" scope has the following in the verbose output:

-> Content-Length: 209
-> 
>> client_id=...&redirect_uri=http%3A%2F%2Flocalhost%3A1410%2F&grant_type=authorization_code&code=8Nma6AxYxYiy3KaSLXgPBoLueoGCcS&client_secret=...

(b) the unsuccessful authentication attempt using "profile,request" scope has:

-> Content-Length: 173
-> 
>> client_id=...&redirect_uri=http%3A%2F%2Flocalhost%3A1410%2F&grant_type=authorization_code&client_secret=...

In the latter case the "code=8Nma6AxYxYiy3KaSLXgPBoLueoGCcS" is absent.

Could this be the source of the problem? My **naive** interpretation is that when I used "profile,request" the code parameter is not being provided in the request.

Best regards,
Andrew.

**Comment by hadley** [comment]
I can't imagine why it would break when you use `profile,request` instead of just `profile`. What happens if you just use `request`?

**Comment by DataWookie** [comment]
Ah, great question!

So, it turns out that if I specify `request` alone then it works. It also works for `history` and `history_lite` in isolation. But, as soon as I pair any two of them together, it breaks.

This works:
```
token <- httr::oauth2.0_token(endpoint, app, scope = "request", cache = FALSE)`
```
but this breaks:
```
token <- httr::oauth2.0_token(endpoint, app, scope = "profile,request", cache = FALSE)
```

But then I went back and **read the documentation properly** and saw that `scope` is "a character vector of scopes to request". After kicking myself in the ass repeatedly, I got this to work:
```
token <- httr::oauth2.0_token(endpoint, app, scope = c("profile", "request"), cache = FALSE)
```

Seems stupid and obvious in retrospect. Sorry for creating noise.
